augur/io/sqlite3.py

```diff
@@ -0,0 +1,137 @@
+        return sqlite3.connect(self.file, **connect_kwargs)
+
+    def execute(self, *args):
+        """Execute under a connection context."""
+        with self.connect() as con:
+            con.execute(*args)
+
+    def executemany(self, *args):
+        """Executemany under a connection context."""
+        # TODO: benchmark performance with cursor.executemany / con.commit at the end
+        with self.connect() as con:
+            con.executemany(*args)
+
+    def create_unique_index(self, table_name:str, column:str):
+        """Create a unique index on a column in a table."""
+        index_name = f'index_{table_name}_{column}'
+        self.execute(f"""
+            CREATE UNIQUE INDEX {sanitize_identifier(index_name)}
+            ON {sanitize_identifier(table_name)} ({sanitize_identifier(column)})
+        """)
+
+    def query_to_file(self, query:str, path:str, header=True, chunksize:int=10000, columns_to_exclude:List[str]=[], **to_csv_kwargs):
+        """Query the database and write results to a tabular file.
+
+        Parameters
+        ----------
+        query
+            SQLite query string.
+        path
+            Path to the output file.
+        header
+            Write out the column names.
+        chunksize
+            Maximum number of rows to load into memory at once.
+        columns_to_exclude
+            Column names to exclude from output.
+        to_csv_kwargs
+            Additional keyword arguments passed to pandas.to_csv().
+        """
+        with self.connect() as con:
+            # Get results in chunks so they are not loaded into memory all at once.
+            df_chunks = pd.read_sql_query(query, con, chunksize=chunksize)
+
+            # Overwrite the file if it exists.
+            mode = 'w'
+
+            for df_chunk in df_chunks:
+                # There is no way to exclude columns in a SQLite query, so it is implemented as an option here.
+                if columns_to_exclude:
+                    for col in columns_to_exclude:
+                        df_chunk.drop(col, axis=1, inplace=True)
+                df_chunk.to_csv(path, mode=mode, header=header, **to_csv_kwargs)
+
+                # For subsequent chunks, use append mode and don't include the header.
+                mode = 'a'
+                header = False
+
+    def delete(self):
+        """Remove the database file if present."""
+        try:
+            os.remove(self.file)
+        except FileNotFoundError:
+            pass
+
+
+def sanitize_identifier(identifier:str):
+    """Sanitize a SQLite identifier."""
+    # Note: we can (and probably should) do more here.
+    # https://stackoverflow.com/a/6701665
+
+    # Escape existing double quotes
+    identifier = identifier.replace('"', '""')
+
+    # Wrap inside double quotes
+    return f'"{identifier}"'
+
+
+def chunked_query_to_csv(con:sqlite3.Connection, query:str, path:str, chunksize:int, header=True, columns_to_exclude:List[str]=[], **to_csv_kwargs):
```

Comment:
I don't see where this function is used, but probably it gets used in the other PR? Separate from its use in this PR, how did you decide to implement this as a function that takes a database connection instead of as a method of the `Sqlite3Database` class? The only difference between this function and the `query_to_file` method is that the function accepts the connection instance as input and the method relies on the connection instance stored in the class. I'm just curious if this split here into a function was based on other design constraints.

Reply:
Oops, this is another artifact of extracting code from the other PR. I think I copied `chunked_query_to_csv` to `query_to_file`, made some changes, and forgot to delete the old copy. Deleted in 4c3f5a63513bb617e5e827f661ca93fe9b24a118.